include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -0,0 +1,370 @@
+        stream << flags << entry;
+        return strstream.str();
+    }
+
+    //!\brief The length of the *str* (traceback symbols are unicode aware)
+    //!\sa https://en.wikipedia.org/wiki/UTF-8 for encoding details
+    static size_t unicode_str_length(std::string const & str) noexcept
+    {
+        size_t length = 0u;
+        for (auto it = str.cbegin(), it_end = str.cend(); it < it_end; ++it, ++length)
+        {
+            uint8_t v = *it;
+            if ((v & 0b11100000) == 0b11000000)
+                ++it;
+            else if ((v & 0b11110000) == 0b11100000)
+                it += 2;
+            else if ((v & 0b11111000) == 0b11110000)
+                it += 3;
+        }
+        return length;
+    }
+
+    //!\brief Format used by seqan3::detail::debug_matrix.
+    struct format_type
+    {
+        //!\brief epsilon symbol (a single symbol)
+        char const * epsilon{};
+        //!\brief column separator symbol (a single symbol)
+        char const * col_sep{};
+        //!\brief row separator symbol (a single symbol)
+        char const * row_sep{};
+        //!\brief row separator symbol (a single symbol)
+        char const * row_col_sep{};
+        //!\brief infinity symbol (a single symbol)
+        char const * inf{};
+    };
+
+    //!\brief The format when printing to a ascii stream.
+    static constexpr format_type csv{" ", ";", "", "", ""};
+    //!\brief The format when printing to a unicode stream.
+    static constexpr format_type unicode{"ε", "║", "═", "╬", "∞"};
+
+public:
+    //!\brief What is the width (number of chars) of an entry. Defaults to #auto_column_width.
+    std::optional<size_t> column_width{std::nullopt};
+
+protected:
+    //!\brief The matrix
+    matrix_t _matrix;
+    //!\brief The first sequence of the sequence alignment.
+    sequence1_t _sequence1;
+    //!\brief The second sequence of the sequence alignment.
+    sequence2_t _sequence2;
+
+    //!\brief Befriend the test case
+    friend struct debug_matrix_print_test;
+};
+
+/*!\name Type deduction guides
+ * \relates seqan3::detail::debug_matrix
+ * \{
+ */
+//!\brief The type deduction guide for the constructor
+//!seqan3::detail::debug_matrix(std::vector<entry_type>, size_t const, size_t const)
+template <typename entry_type>
+debug_matrix(std::vector<entry_type>, size_t const, const size_t)
+    -> debug_matrix<row_wise_matrix<entry_type>>;
+
+//!\brief The type deduction guide for the constructor
+//!seqan3::detail::debug_matrix(std::vector<entry_type>, size_t const, size_t const, sequence1_t, sequence2_t)
+template <typename entry_type, typename sequence1_t, typename sequence2_t>
+debug_matrix(std::vector<entry_type>, size_t const, const size_t, sequence1_t &&, sequence2_t &&)
+   -> debug_matrix<row_wise_matrix<entry_type>, sequence1_t, sequence2_t>;
+
+
+//!\brief The type deduction guide for the constructor seqan3::detail::debug_matrix(matrix_t)
+template <Matrix matrix_t>
+debug_matrix(matrix_t &&)
```

Comment:
do you need this type deduction guide? I thought the compiler can deduce this kind on its own

Reply:
No it is needed.